internal/destination/destination.go

```diff
@@ -131,42 +132,79 @@ func (d *Destination) Open(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("connect to NATS: %w", err)
 	}
-
-	d.writer, err = NewWriter(WriterParams{
-		Conn:          conn,
-		Subject:       d.config.Subject,
-		RetryWait:     d.config.RetryWait,
-		RetryAttempts: d.config.RetryAttempts,
-	})
-	if err != nil {
-		return fmt.Errorf("init jetstream writer: %w", err)
-	}
+	d.nc = conn
 
 	// Async handlers & callbacks
 	conn.SetErrorHandler(internal.ErrorHandlerCallback(ctx))
-	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {}))
-	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {}))
+	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {
+		d.writer.stopWrites()
+	}))
+	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {
+		d.writer, err = NewWriter(writerParams{
+			nc:            d.nc,
+			subject:       d.config.Subject,
+			retryWait:     d.config.RetryWait,
+			retryAttempts: d.config.RetryAttempts,
+		})
+		d.writer.startWrites()
+	}))
 	conn.SetClosedHandler(internal.ClosedCallback(ctx))
 	conn.SetDiscoveredServersHandler(internal.DiscoveredServersCallback(ctx))
 
+	d.writer, err = NewWriter(writerParams{
+		nc:            d.nc,
+		subject:       d.config.Subject,
+		retryWait:     d.config.RetryWait,
+		retryAttempts: d.config.RetryAttempts,
+	})
+	if err != nil {
+		return fmt.Errorf("init jetstream writer: %w", err)
+	}
+
 	return nil
 }
 
 // Write writes a record into a Destination.
-func (d *Destination) Write(_ context.Context, records []sdk.Record) (int, error) {
-	for i, record := range records {
-		if err := d.writer.Write(record); err != nil {
-			return i, fmt.Errorf("write: %w", err)
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	recorded := 0
+	for _, record := range records {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			// if the context error has not a cause, err will have a null value
+			if err == nil {
+				err = context.Canceled
+			}
+			sdk.Logger(ctx).Debug().
+				Int("record total", len(records)).
+				Int("record recorded", recorded).
+				Err(err).
+				Msg("write stopped by context before having all records recorded")
+
+			return recorded, err
+		default:
+			if err := d.writer.Write(record); err != nil {
+				sdk.Logger(ctx).Debug().
+					Int("record total", len(records)).
+					Int("record recorded", recorded).
+					Err(err).
+					Send()
```

Comment:
Wait, so now we are just logging the error and continue writing the next record? And we even increase `recorded` so it will seem like the record was successfully written 😕 This could cause records to go missing.